api/v1alpha3/docluster_webhook.go

```diff
@@ -0,0 +1,78 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha3
+
+import (
+	"fmt"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	runtime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	ctrl "sigs.k8s.io/controller-runtime"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+// log is for logging in this package.
+var _ = logf.Log.WithName("docluster-resource")
+
+// +kubebuilder:webhook:verbs=create;update,path=/validate-infrastructure-cluster-x-k8s-io-v1alpha3-docluster,mutating=false,failurePolicy=fail,matchPolicy=Equivalent,groups=infrastructure.cluster.x-k8s.io,resources=doclusters,versions=v1alpha3,name=validation.docluster.infrastructure.cluster.x-k8s.io,sideEffects=None
+// +kubebuilder:webhook:verbs=create;update,path=/mutate-infrastructure-cluster-x-k8s-io-v1alpha3-docluster,mutating=true,failurePolicy=fail,matchPolicy=Equivalent,groups=infrastructure.cluster.x-k8s.io,resources=doclusters,versions=v1alpha3,name=default.docluster.infrastructure.cluster.x-k8s.io,sideEffects=None
+
+var (
+	_ webhook.Defaulter = &DOCluster{}
+	_ webhook.Validator = &DOCluster{}
+)
+
+func (r *DOCluster) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type
+func (r *DOCluster) Default() {}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type.
+func (r *DOCluster) ValidateCreate() error {
+	return nil
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (r *DOCluster) ValidateUpdate(old runtime.Object) error {
+	var allErrs field.ErrorList
+
+	oldDOCluster, ok := old.(*DOCluster)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected an DOCluster but got a %T", old))
+	}
+
+	if r.Spec.Region != oldDOCluster.Spec.Region {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("spec", "region"), r.Spec.Region, "field is immutable"))
+	}
+
+	if allErrs != nil {
```

Comment:
Should we also prevent changes to the `controlPlaneEndpoint` field?

It consists of the IP address of the LB which is immutable once set, and the LB port which I don't _think_ we allow updating right now?

(I did wonder at some point why those were spec fields in the first place as opposed to status ones, but maybe I'm missing some context here...)

Reply:
+1 

> (I did wonder at some point why those were spec fields in the first place as opposed to status ones, but maybe I'm missing some context here...)

It's required by capi to identify the endpoint used to connect to the target’s cluster apiserver.
https://cluster-api.sigs.k8s.io/developer/architecture/controllers/cluster.html#infrastructure-provider

Reply:
Right, what why is the endppint part of the spec? AFAIU, CAPDO does not expect the user/operator to set or modify the field -- it's always the controller that does it. In that sense, a status field seems better?

(Definitely a side discussion since this isn't something we can change without CAPI approval.)